File: Modulo 4/Esercizio/dotnet/CalculatorApp/Calculator.cs
using System;
using System.Collections.Generic;

namespace CalculatorApp
{
    public class Calculator
    {
        private List<string> history;

        public Calculator()
        {
            history = new List<string>();
        }

        public double Add(double a, double b)
        {
            double result = a + b;
            history.Add($"{a} + {b} = {result}");
            return result;
        }

        public double Subtract(double a, double b)
        {
            double result = a - b;
            history.Add($"{a} - {b} = {result}");
            return result;
        }

        public double Multiply(double a, double b)
        {
            double result = a * b;
            history.Add($"{a} * {b} = {result}");
            return result;
        }

        public double Divide(double a, double b)
        {
            if (b == 0)
            {
                throw new DivideByZeroException("Cannot divide by zero.");
            }
            double result = a / b;
            history.Add($"{a} / {b} = {result}");
            return result;
        }

        public double Power(double baseNum, int exponent)
        {
            double result = 1;
            for (int i = 0; i < exponent; i++)
            {
                result *= baseNum;
            }
            history.Add($"{baseNum} ^ {exponent} = {result}");
            return result;
        }

        public List<string> GetHistory()
        {
            return history;
        }

        public void ClearHistory()
        {
            history.Clear();
        }

        public double Percentage(double value, double percent)
        {
            double result = (value * percent) / 100;
            history.Add($"{percent}% of {value} = {result}");
            return result;
        }
    }
}

File: Modulo 4/Esercizio/typescript/Calculator.ts
export class Calculator {
    private history: string[];

    constructor() {
        this.history = [];
    }

    public add(a: number, b: number): number {
        let result = a + b;
        this.history.push(`${a} + ${b} = ${result}`);
        return result;
    }

    public subtract(a: number, b: number): number {
        let result = a - b;
        this.history.push(`${a} - ${b} = ${result}`);
        return result;
    }

    public multiply(a: number, b: number): number {
        let result = a * b;
        this.history.push(`${a} * ${b} = ${result}`);
        return result;
    }

    public divide(a: number, b: number): number {
        if (b === 0) {
            throw new Error("Cannot divide by zero.");
        }
        let result = a / b;
        this.history.push(`${a} / ${b} = ${result}`);
        return result;
    }

    public power(baseNum: number, exponent: number): number {
        let result = 1;
        for (let i = 0; i < exponent; i++) {
            result *= baseNum;
        }
        this.history.push(`${baseNum} ^ ${exponent} = ${result}`);
        return result;
    }

    public getHistory(): string[] {
        return this.history;
    }

    public clearHistory(): void {
        this.history = [];
    }

    public percentage(value: number, percent: number): number {
        let result = (value * percent) / 100;
        this.history.push(`${percent}% of ${value} = ${result}`);
        return result;
    }
}
